### DISCUSSION TITLE
How to run Docker in interactive mode

### DISCUSSION THREAD

**Discussion by NishkarshRaj** [community/community#25509] [Actions]
<p>I have used Actions to add a .cpp file from repo to Docker Image and executed it successfully after installation of g++ in the Docker container.<br>
The Dockerfile is automatically executed in Actions but I am unable to fetch the output after using ./a.out because the Docker container is not running in interactive mode (-it).</p>
<p>Please help, Link: <a href="https://github.com/NishkarshRaj/Computer-Graphics" rel="nofollow noopener">https://github.com/NishkarshRaj/Computer-Graphics</a></p>

**Comment by Yanjingzhu** [comment]
<p>I checked your docker action. You have two CMD instructions in your dockerfile. According to <a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.docker.com%2Fengine%2Freference%2Fbuilder%2F%23cmd&amp;data=02%7C01%7Cv-yjingz%40microsoft.com%7C2bc1bfdec41548857fde08d7839a28b1%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637122570441827676&amp;sdata=Z5eoKljatu%2B22jSfB0zQ6ZOi%2Fu68o3ncsm4GFUfVmBE%3D&amp;reserved=0" rel="nofollow noopener">docker document</a>: There can only be one CMD instruction in a Dockerfile. If you list more than one CMD then only the last CMD will take effect.</p>
<p><img src="https://user-images.githubusercontent.com/24806478/181091654-4c067e03-8e40-493d-8685-1c11a50ed951.png" alt="2CMD.png" data-base62-sha1="wsq3zsghgSMP97nK0k58of4Xqan" width="586" height="332"></p>
<p>So only <em>echo “Dockerfile succesfully executed. Now running the Shell Script”</em> is executed . <strong>script1.sh</strong> will not.</p>
<p>Could you please correct your dockerfile and make the script1.sh be executed?</p>
<p>Note: When execute script1.s,. you could use <strong>ls</strong> command to check whether a.out is in the directory.</p>

**Comment by NishkarshRaj** [comment]
<p>Thanks for replying.</p>
<p>I have made two changes based on your suggestion:</p>
<ol>
<li>
<p>I have removed the last CMD command</p>
</li>
<li>
<p>I have added ls command in the start of the shell script.</p>
</li>
</ol>
<p>Now, the problem is, Actions is not able to find the script file. I have been facing this problem for long but then I thought maybe it is not able to pick the path so I hardcoded the path and set my WORKDIR of docker as that path.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/76a242431a25c589b20b0885ada2b4761c919874.png" data-download-href="/uploads/short-url/gVtYJ4kLWlUXDvcIyMZqF55MGm8.png?dl=1" title="Screenshot (636).png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/35298207/181091655-106aefbe-af6c-446f-ba01-4bf8cc0b403b.png" alt="Screenshot (636).png" data-base62-sha1="gVtYJ4kLWlUXDvcIyMZqF55MGm8" width="690" height="411" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/76a242431a25c589b20b0885ada2b4761c919874_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot (636).png</span><span class="informations">1069×637 38.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by NishkarshRaj** [comment]
<p>I have also added the pwd and ls command, and they confirm that I am in the needed directory and my script file is also present there.</p>
<p>I don’t know why it states the file is not present on execution <img src="https://user-images.githubusercontent.com/35298207/181091658-10ded348-4807-4f9b-bf7a-ad66bf63bf7f.png" alt="Screenshot (637).png" data-base62-sha1="sh4D3Yad1CRbD4qUikpfqAYmPaQ" width="377" height="187"></p>

**Comment by chrispat** [comment]
<p>In your example below you have set your WORKDIR to /nish and then are trying to run ./nish/script1.sh which would say the script is at /nish/nish/script1.sh.  Try changing your CMD to ./script1.sh.</p>

**Comment by NishkarshRaj** [comment]
<p>That will not work sir. I am using Ubuntu, so any path I specify wrt root (/), it will take as absolute path rather than the relative path. Thus my script will be on /nish/script1.sh only rather than /nish/nish/script1.sh. Anyways, I have made the changes you suggested and still I am getting the same error, the script is not found as can be seen in attached image. <img src="https://user-images.githubusercontent.com/35298207/181091662-32c907a8-ead5-49b7-9b15-7e2eaffad22b.png" alt="Screenshot (638).png" data-base62-sha1="j1UFvn6LW2SK2TfECday2hL6vIR" width="478" height="343"></p>

**Comment by chrispat** [comment]
<p>There are a couple of interesting things here now that I have looked at your code.</p>
<p>First, I am curious of your scenario and why you are wrapping up a single run of gcc in an action for a single file.  That does not seem very resuable.</p>
<p>Second, as per the documentation for container actions once your action has been built it is run and we pass --workdir as part of the docker run so the working directory is always the github workspace where the code is.  This is so actions can be reusable and have a consistent contract.</p>
<p>Having made a couple of small changes to account for that the action is still going to fail becuase there is no display inside the container <a href="https://github.com/chrispat/Computer-Graphics/commit/98f5c849ce60a8d21fa4815e06f44fba8c36b292/checks?check_suite_id=365204691#step:3:953" rel="nofollow noopener">https://github.com/chrispat/Computer-Graphics/commit/98f5c849ce60a8d21fa4815e06f44fba8c36b292/checks?check_suite_id=365204691#step:3:953</a>.</p>

**Comment by NishkarshRaj** [comment]
<p>Thanks alot sir for replying.</p>
<p>I know my Actions Dockerfile is not general, actually I posted this thread just to check whether the code will generate any output in the container or not and then I would later would have worked to make it generic for all source commits.</p>
<p>So by your reply and link attached, can we just state that the Docker container in Actions cannot display the output of ./a.out in interactive mode?</p>

**Comment by NishkarshRaj** [comment, marked as answer]
<p>Thanks alot sir.</p>
<p>The biggest realization from your reply was:</p>
<ol>
<li>
<p>/github/workflows or root of the repository is the workflow for GitHub Actions containers.</p>
</li>
<li>
<p>a.out file is accessible after doing g++ on “action-a/Hut.cpp” file as seen in attached image but the container cannot display the output.</p>
</li>
</ol>
<p>My actions file is now running properly and not giving any errors.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/a/abd1361d515595d254976ad8856337a780bd9c27.png" data-download-href="/uploads/short-url/ovXSBGxb2pCX6eKP1LUHACamLMr.png?dl=1" title="Screenshot (650).png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/35298207/181091664-1eb62e91-6d35-4030-89fc-73c5f80981d8.png" alt="Screenshot (650).png" data-base62-sha1="ovXSBGxb2pCX6eKP1LUHACamLMr" width="497" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/a/abd1361d515595d254976ad8856337a780bd9c27_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screenshot (650).png</span><span class="informations">576×579 15.7 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
